### PR TITLE
feat: Set to Function

### PR DESCRIPTION
**Pull request by madvorak**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Set/ToFunction.lean [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Martin Dvorak
+-/
+import Mathlib.Data.Set.Defs
+import Mathlib.Logic.ExistsUnique
+
+/-!
+# Set to function
+
+This file provides API for sets that define functions.
```
Comment (vihdzp) [inline Mathlib/Data/Set/ToFunction.lean:12]: I'd add to this docstring that these notions mimic the standard definition of a function in ZFC.

File: Mathlib/Data/Set/ToFunction.lean [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+## Main definitions
+- `Set.asPartialFun` converts `Set (α × α)` to `α → Option α`.
+- `Set.asFun` converts `Set (α × α)` to `α → α` if possible.
+
```
Comment (vihdzp) [inline Mathlib/Data/Set/ToFunction.lean:17]:
```suggestion

- `Set.asPartialFun` converts `Set (α × α)` to `α → Option α`.
- `Set.asFun` converts `Set (α × α)` to `α → α` if possible.
```

File: Mathlib/Data/Set/ToFunction.lean [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+-/
+
+variable {α : Type*}
+
+namespace Set
+
+section set_as_partial_function
+
+/-- Does given set on `α × α` represent a partial function? -/
```
Comment (vihdzp) [inline Mathlib/Data/Set/ToFunction.lean:26]: This isn't a very good docstring, as it essentially just clarifies the name `isPartialFun` without explaining what it *means*. I'd write something like
```suggestion
/-- A set `s : Set (α × α)` represents a partial function when for every `x : α` there's at most
one `y : α` with `(x, y) ∈ s`. -/
```

File: Mathlib/Data/Set/ToFunction.lean [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+def isPartialFun (X : Set (α × α)) : Prop :=
+  ∀ x y z : α, (x, y) ∈ X ∧ (x, z) ∈ X → y = z
```
Comment (vihdzp) [inline Mathlib/Data/Set/ToFunction.lean:28]: You can write this as
```suggestion
  ∀ x : α, {y : α | (x, y) ∈ X}.Subsingleton
```
I believe you can even get rid of the type annotations if you want.

File: Mathlib/Data/Set/ToFunction.lean [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+-/
+
+variable {α : Type*}
+
+namespace Set
+
+section set_as_partial_function
+
+/-- Does given set on `α × α` represent a partial function? -/
+def isPartialFun (X : Set (α × α)) : Prop :=
```
Comment (vihdzp) [inline Mathlib/Data/Set/ToFunction.lean:27]: Capital letters are usually reserved for types; you should use something like `x` or `s` instead. This applies throughout the file.

Also, this should be named `IsPartialFun`.

File: Mathlib/Data/Set/ToFunction.lean [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  ∀ x y z : α, (x, y) ∈ X ∧ (x, z) ∈ X → y = z
+
+open Classical in
+/-- Use given set on `α × α` as a partial function. -/
```
Comment (vihdzp) [inline Mathlib/Data/Set/ToFunction.lean:31]: Again, this docstring doesn't clarify much. You can write something like
```suggestion
/-- Turns `s : Set (α × α)` into a partial function. Each `x : α` is mapped to an arbitrary `y : α`
with `(x, y) ∈ s`, or to `none` if none exists. -/
```

File: Mathlib/Data/Set/ToFunction.lean [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+noncomputable def asPartialFun (X : Set (α × α)) : α → Option α :=
+  fun a : α => if hb : ∃ b, (a, b) ∈ X then hb.choose else none
+
+theorem αsPartialFun_eq {X : Set (α × α)} (hX : isPartialFun X) {a b : α} (hab : (a, b) ∈ X) :
```
Comment (vihdzp) [inline Mathlib/Data/Set/ToFunction.lean:35]:
```suggestion
theorem asPartialFun_eq {X : Set (α × α)} (hX : isPartialFun X) {a b : α} (hab : (a, b) ∈ X) :
```

File: Mathlib/Data/Set/ToFunction.lean [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+- `Set.asPartialFun` converts `Set (α × α)` to `α → Option α`.
+- `Set.asFun` converts `Set (α × α)` to `α → α` if possible.
+
+-/
+
+variable {α : Type*}
+
+namespace Set
+
+section set_as_partial_function
```
Comment (vihdzp) [inline Mathlib/Data/Set/ToFunction.lean:24]: No need for such a long section name
```suggestion
section partial
```

File: Mathlib/Data/Set/ToFunction.lean [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+  fun a : α => if hb : ∃ b, (a, b) ∈ X then hb.choose else none
+
+theorem αsPartialFun_eq {X : Set (α × α)} (hX : isPartialFun X) {a b : α} (hab : (a, b) ∈ X) :
+    asPartialFun X a = b := by
+  have hba : ∃ b, (a, b) ∈ X := ⟨b, hab⟩
+  simpa [asPartialFun, hba] using hX _ _ _ ⟨hba.choose_spec, hab⟩
+
+end set_as_partial_function
+
+section set_as_total_function
```
Comment (vihdzp) [inline Mathlib/Data/Set/ToFunction.lean:42]:
```suggestion
section total
```

File: Mathlib/Data/Set/ToFunction.lean [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+/-- Does given set on `α × α` represent a total function? -/
```
Comment (vihdzp) [inline Mathlib/Data/Set/ToFunction.lean:44]:
```suggestion
/-- A set `s : Set (α × α)` represents a total function when for every `x : α` there's exactly one
`y : α` with `(x, y) ∈ s`. -/
```

File: Mathlib/Data/Set/ToFunction.lean [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+def isFun (X : Set (α × α)) : Prop :=
+  ∀ x : α, ∃! y : α, (x, y) ∈ X
+
+theorem isPartialFun_of_isFun {X : Set (α × α)} (hX : isFun X) : isPartialFun X := by
```
Comment (vihdzp) [inline Mathlib/Data/Set/ToFunction.lean:48]: You can employ dot notation here
```suggestion
theorem isFun.isPartialFun {X : Set (α × α)} (hX : isFun X) : isPartialFun X := by
```

File: Mathlib/Data/Set/ToFunction.lean [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    asPartialFun X a = b := by
+  have hba : ∃ b, (a, b) ∈ X := ⟨b, hab⟩
+  simpa [asPartialFun, hba] using hX _ _ _ ⟨hba.choose_spec, hab⟩
+
+end set_as_partial_function
+
+section set_as_total_function
+
+/-- Does given set on `α × α` represent a total function? -/
+def isFun (X : Set (α × α)) : Prop :=
```
Comment (vihdzp) [inline Mathlib/Data/Set/ToFunction.lean:45]: Should be named `IsFun`.

File: Mathlib/Data/Set/ToFunction.lean [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+  intro x y z ⟨hxy, hxz⟩
+  have hy := (hX x).choose_spec.2 y hxy
+  have hz := (hX x).choose_spec.2 z hxz
+  exact hy.trans hz.symm
+
+/-- Use given set on `α × α` as a total function. -/
+noncomputable def asFun {X : Set (α × α)} (hX : isFun X) : α → α :=
+  fun a : α => (hX a).choose
+
+theorem αsFun_eq {X : Set (α × α)} (hX : isFun X) {a b : α} (hab : (a, b) ∈ X) :
```
Comment (vihdzp) [inline Mathlib/Data/Set/ToFunction.lean:58]:
```suggestion
theorem asFun_eq {X : Set (α × α)} (hX : isFun X) {a b : α} (hab : (a, b) ∈ X) :
```

File: Mathlib/Data/Set/ToFunction.lean [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+def isFun (X : Set (α × α)) : Prop :=
+  ∀ x : α, ∃! y : α, (x, y) ∈ X
+
+theorem isPartialFun_of_isFun {X : Set (α × α)} (hX : isFun X) : isPartialFun X := by
+  intro x y z ⟨hxy, hxz⟩
+  have hy := (hX x).choose_spec.2 y hxy
+  have hz := (hX x).choose_spec.2 z hxz
+  exact hy.trans hz.symm
+
+/-- Use given set on `α × α` as a total function. -/
```
Comment (vihdzp) [inline Mathlib/Data/Set/ToFunction.lean:54]:
```suggestion
/-- Turns `s : Set (α × α)` into a total function. Each `x : α` is mapped to the unique `y : α` with
`(x, y) ∈ s`, or to `none` if none exists. -/
```

File: Mathlib/Data/Set/ToFunction.lean [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+import Mathlib.Data.Set.Defs
+import Mathlib.Logic.ExistsUnique
+
+/-!
+# Set to function
+
+This file provides API for sets that define functions.
+
+## Main definitions
+- `Set.asPartialFun` converts `Set (α × α)` to `α → Option α`.
```
Comment (vihdzp) [inline Mathlib/Data/Set/ToFunction.lean:15]: Why not convert `Set (α × β)` into `α → Option β` / `α → β` instead?

**Review by vihdzp — CHANGES_REQUESTED** [review]
I'm very skeptical that any of this is particularly useful. Lean's type theory doesn't constructively satisfy the principle of unique choice, and as such, it's impossible to turn `Set (α × β)` into a function `α → β` constructively.

In general, if you have `s : Set (α × β)` and can give a precise correspondence between the elements of `α` and the elements of `β`, then you can just use said correspondence to build a function. If you don't have such a correspondence, but you still need to create a function, it's probably better to just employ choice explicitly. In this case, the predicate `isFun` in this PR might potentially be too strong of an assumption, depending on what precisely you're doing - you might simply need `∀ x : α, ∃ y : β, (x, y) ∈ s`.

Do you have a particular need for these functions in mind? If so, I've left a bunch of comments anyways.

File: Mathlib/Data/Set/ToFunction.lean (hunk 2 above)
Comment (madvorak) [reply Mathlib/Data/Set/ToFunction.lean:17]: I am confused. What did you change here?

File: Mathlib/Data/Set/ToFunction.lean (hunk 15 above)
Comment (madvorak) [reply Mathlib/Data/Set/ToFunction.lean:15]: Good idea! Thanks!

File: Mathlib/Data/Set/ToFunction.lean (hunk 2 above)
Comment (vihdzp) [reply Mathlib/Data/Set/ToFunction.lean:17]: I added a blank line above and removed a blank line below.

File: Mathlib/Data/Set/ToFunction.lean [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+end set_as_partial_function
+
+section set_as_total_function
+
+/-- A set `s : Set (α × α)` represents a total function when for every `x : α` there's exactly one
+`y : β` such that `(x, y) ∈ s`. -/
+def IsFun (s : Set (α × β)) : Prop :=
+  ∀ x : α, ∃! y : β, (x, y) ∈ s
+
+theorem isFun.isPartialFun {s : Set (α × β)} (hs : IsFun s) : IsPartialFun s :=
```
Comment (vihdzp) [inline Mathlib/Data/Set/ToFunction.lean:58]:
```suggestion
theorem IsFun.isPartialFun {s : Set (α × β)} (hs : IsFun s) : IsPartialFun s :=
```

File: Mathlib/Data/Set/ToFunction.lean [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+  ((hs a).choose_spec.2 b hab).symm
+
+end set_as_total_function
+
+end Set
+
+/-- Convert `Rel` to a `Set` (like uncurry). -/
+def Rel.toSet (R : Rel α β) : Set (α × β) := fun ⟨a, b⟩ => R a b
+
+theorem Function.graph_is_fun (f : α → β) : f.graph.toSet.IsFun :=
```
Comment (vihdzp) [inline Mathlib/Data/Set/ToFunction.lean:77]:
```suggestion
theorem Function.graph_isFun (f : α → β) : f.graph.toSet.IsFun :=
```

File: Mathlib/Data/Set/ToFunction.lean [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+  fun x => ⟨f x, rfl, fun _ => Eq.symm⟩
+
+theorem Function.graph_toSet_asFun (f : α → β) : Set.asFun f.graph_is_fun = f :=
```
Comment (vihdzp) [inline Mathlib/Data/Set/ToFunction.lean:80]:
```suggestion
@[simp]
theorem Function.asFun_graph (f : α → β) : Set.asFun f.graph_is_fun = f :=
```

**Review by vihdzp — COMMENTED** [review]
More suggestions. I still really want to know what you plan to use this for.

File: Mathlib/Data/Set/ToFunction.lean (hunk 8 above)
Comment (madvorak) [reply Mathlib/Data/Set/ToFunction.lean:31]: I don't like the idea of having a section name identical to a keyword, even if it is allowed.

Comment (Rida-Hamadani) [reply Mathlib/Data/Set/ToFunction.lean:31]: As far as I know, there are no snake case sections in mathlib, and I support consistency. Also, I suggest using docs as description of the section's role. In this case, adding a 
```
/-! ### Sets as Partial Functions -/
```
before the section.

File: Mathlib/Data/Set/ToFunction.lean [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+/-- A set `s : Set (α × α)` represents a partial function if for every `x : α` there is at most one
+`y : β` such that `(x, y) ∈ s`. -/
+def IsPartialFun (s : Set (α × β)) : Prop :=
+  ∀ x : α, { y : β | (x, y) ∈ s }.Subsingleton
+
+open Classical in
+/-- Use given set on `α × α` as a partial function. Each `x : α` is mapped to the unique `y : β`
+such that `(x, y) ∈ s`, or to `none` if none exists.  -/
+noncomputable def asPartialFun (s : Set (α × β)) : α → Option β :=
+  fun a : α => if hb : ∃ b, (a, b) ∈ s then hb.choose else none
```
Comment (urkud) [inline Mathlib/Data/Set/ToFunction.lean:42]: Did you consider `PFun` here?

Comment (madvorak) [reply Mathlib/Data/Set/ToFunction.lean:42]: You mean as a name for the noncomputable definition?

**Comment by madvorak** [comment]
> More suggestions. I still really want to know what you plan to use this for.

The use case I had in mind no longer exists.

Nevertheless, other users might search for it, because given relationship between functions and sets are a standard part of the undergraduate curriculum (perhaps immediately in the first week for pure math students).

**Comment by vihdzp** [comment]
The relationship between functions and sets arguably already exists as [ZFSet.IsFunc](https://leanprover-community.github.io/mathlib4_docs/Mathlib/SetTheory/ZFC/Basic.html#ZFSet.IsFunc).

Defining functions as sets is a technical necessity of ZFC which for whatever reason seems to have reached the high school curriculum. In type theory, functions are a primitive, and this encoding is not only unnecessary but strictly less convenient, at least in my view.

**Comment by madvorak** [comment]
Just one question; what is the type-theoretic way to say that a partial function extends another partial function?

**Comment by vihdzp** [comment]
I'm not sure if we have something like this in Mathlib, but given `f g : α → Option β` you could define a relation `f ⪯ g` as `∀ x, (f x).isSome → f x = g x`.

**Comment by madvorak** [comment]
So, shall we close the PR now that we know it is redundant?

**Comment by vihdzp** [comment]
I vote close.

**Comment by madvorak** [comment]
Closing as unnecessary.

**Pull request closed without merging**